### PR TITLE
[WIP] Remove MMU-hack, and always turn on MMU.

### PR DESCRIPTION
**Pull request by magumagu**
Making the assumption that games never cause DSI or ISI exceptions doesn't provide much performance benefit, and keeping around the fake memory region makes improving our MMU implementation (e.g. adding a correct BAT implementation) difficult, and JIT modifications in general more complicated.

As a bonus, this makes it harder for games to crash Dolphin.

Probably doesn't make sense to start seriously benchmarking/testing this until Fiora's latest batch of changes are merged; there are a couple of important changes in there which affect the performance of non-MMU games in MMU mode.  I want to start the discussion now, though, to make sure there isn't some other reason I'm missing to keep around MMU-disabled mode.

This patch doesn't remove all the code which checks `ADDR_MASK_MEM1` (0x20000000) because we still make assumptions about that bit in addresses even if we don't physically map a fake-vmem region; fixing that mess will probably be a followup to this patch.

---

This change is [Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/1831)

### CODE REVIEW

File: Source/Core/Core/HW/MemmapFunctions.cpp [hunk 1]
```diff
@@ -348,8 +324,7 @@ u32 Read_Opcode(u32 _Address)
 		return 0x00000000;
 	}
 
-	if (SConfig::GetInstance().m_LocalCoreStartupParameter.bMMU &&
-		(_Address & ADDR_MASK_MEM1))
+	if ((_Address & ADDR_MASK_MEM1))
```
Comment (BhaaLseN) [inline Source/Core/Core/HW/MemmapFunctions.cpp]: You could get rid of the extra brackets.

File: Source/Core/Core/PowerPC/JitInterface.cpp [hunk 2]
```diff
@@ -209,7 +203,7 @@ namespace JitInterface
 
 	u32 ReadOpcodeJIT(u32 _Address)
 	{
-		if (bMMU && !bFakeVMEM && (_Address & Memory::ADDR_MASK_MEM1))
+		if ((_Address & Memory::ADDR_MASK_MEM1))
```
Comment (BhaaLseN) [inline Source/Core/Core/PowerPC/JitInterface.cpp]: Also down here.

File: Source/Core/Core/HW/MemmapFunctions.cpp (hunk 1 above)
Comment (magumagu) [inline Source/Core/Core/HW/MemmapFunctions.cpp]: IIRC clang warns if you don't put in the extra parens?  Or maybe I'm misremembering.

**Comment by skidau** [comment]
This PR contains straightforward code - removes the MMU option and removes the MMU speed hack code.

In my opinion, we're going to need some sort of DSI/ISI detection and speed hack to ensure that no games take a significant speed hit.  I have attempted to write code that detects if the BATs and other MMU regs have been modified from their standard values, but the code detected the situation too late to enable/disable the MMU.

All games install a DSI/ISI exception handler which makes it hard to detect the games that require the MMU vs ones that don't.

One other idea I have is that the MMU speed hack games might use a standard VMEM allocation function which could be detected using a Dolphin function signature.

File: Source/Core/Core/HW/MemmapFunctions.cpp (hunk 1 above)
Comment (BhaaLseN) [inline Source/Core/Core/HW/MemmapFunctions.cpp]: Does it? We don't do it in other places either iirc.

**Comment by Xcedf** [comment]
Didn't do much of testing, but some non-Mmu games became a little slower when mmu-speedhack was removed, can't even imagine how slow will it be with Always mmu

**Comment by phire** [comment]
There are a few potential optimisations that can be made based on the assumption of no DSI exceptions. Things like re-ordering of writes or merging writes together.

But for cleaner code this might be a good idea.

File: Source/Core/Core/HW/MemmapFunctions.cpp (hunk 1 above)
Comment (lioncash) [inline Source/Core/Core/HW/MemmapFunctions.cpp]: @magumagu I believe that's only when there's a bitwise OR present. e.g. Something like `(val & 0xFFFF | val >> 16)`

**Comment by phire** [comment]
Another downside of this idea is that non-fastmem will suddenly get a lot slower.

**Comment by degasus** [comment]
Who cares about the performance of non-fastmem performance? @phire We could just rethrow real segfaults as sigaborts and gdb will be happy

**Comment by ZephyrSurfer** [comment]
Right, I don't know exactly how to argue with people or state my views correctly
But what about the already lack of error handling where Dolphin will just plain crash Turning on MMU hides all of these errors for that alone I don't like the way this is going :(

**Comment by phire** [comment]
Dolphin is crashing due to the lack of MMU code. 
With MMU enabled that error handling is enabled, so it either works correctly, or it sends an DSI to the game.

**Comment by magumagu** [comment]
@ZephyrSurfer Without MMU, if a game does an invalid read, we PanicAlert or Dolphin crashes; with MMU, the game crashes, and we get a nice crash report (something like http://imgur.com/1GCRJoU).  This is an improvement.

File: Source/Core/Core/PowerPC/Jit64/Jit.cpp [hunk 3]
```diff
@@ -177,7 +177,7 @@ void Jit64::Init()
 
 	jo.optimizeGatherPipe = true;
 	jo.accurateSinglePrecision = true;
-	js.memcheck = SConfig::GetInstance().m_LocalCoreStartupParameter.bMMU;
+	js.memcheck = true;
```
Comment (myownfriend) [inline Source/Core/Core/PowerPC/Jit64/Jit.cpp]: You actually don't need this variable at all anymore.

In the rest of the file, it's used only a few more times:

trampolines.Init(js.memcheck ? TRAMPOLINE_CODE_SIZE_MMU : TRAMPOLINE_CODE_SIZE);

farcode.Init(js.memcheck ? FARCODE_SIZE_MMU : FARCODE_SIZE);

if (js.memcheck && (opinfo->flags & FL_LOADSTORE))

Those can become

trampolines.Init(TRAMPOLINE_CODE_SIZE_MMU);

farcode.Init(FARCODE_SIZE_MMU);

if (opinfo->flags & FL_LOADSTORE)

File: Source/Core/Core/PowerPC/Jit64IL/JitIL.cpp [hunk 4]
```diff
@@ -247,7 +247,7 @@ void JitIL::Init()
 
 	jo.optimizeGatherPipe = true;
 	jo.accurateSinglePrecision = false;
-	js.memcheck = SConfig::GetInstance().m_LocalCoreStartupParameter.bMMU;
+	js.memcheck = true;
```
Comment (myownfriend) [inline Source/Core/Core/PowerPC/Jit64IL/JitIL.cpp]: Same thing is true here.

**Comment by FioraAeterna** [comment]
Isn't there a fake_vmem "jit icache" used in jitasm and the block cache that you could remove too?

**Comment by myownfriend** [comment]
I'm not super familiar with Dolphin's code but, after doing a grep for "memcheck", the line I commented on above seems to be the only time it's actually set to anything. Is memcheck only used to check if MMU emulation is enabled? If that's the case, then there are way more areas where code can be removed.

**Comment by magumagu** [comment]
@myownfriend There's a bunch of other code you're not finding that depends on the memcheck setting... I didn't remove it all because it would make this difficult to rebase on top of master.

@FioraAeterna The jit cache's "vmem" shortcut is conceptually related, but the code isn't directly affected by the removal of the fake vmem region.  It's just doing some nonsensical masking to reduce the size of its map from effective addresses to jit blocks.

**Comment by myownfriend** [comment]
@magumagu I found plenty of code that depends on the setting, just nothing that changes what value its storing. It's always equals true.

**Comment by magumagu** [comment]
Yes, that's correct.

**Comment by myownfriend** [comment]
@magumagu so all the code that I found will likely be removed or changed later on?

**Comment by magumagu** [comment]
Yes.

**Comment by myownfriend** [comment]
Okay, awesome. Was just making sure I wasn't stupid lol

**Comment by magumagu** [comment]
Update: keeping around fake-vmem in PR #1882 wasn't as hard as I thought it would be, so this is less of a priority for me.  I'll see where we stand once dynamic-bat is merged.

**Comment by Fallcrest** [comment]
I know @magumagu recently has been busy, but what is the status of this now that the Dynamic BAT has been implemented? Does it work as is?

**Comment by JMC47** [comment]
While Enable MMU has no performance penalty on x86-64, JITARM takes a performance hit.  We have to wait for now.

**Comment by Fallcrest** [comment]
@JM47, okay, thanks. I didn't know that.

**Comment by degasus** [comment]
@Fallcrest There is no need to rush here, as the harder parts are already done within dynamic_bat ;) Dropping the ini and removing dead code should be easy.

**Comment by delroth** [comment]
https://github.com/dolphin-emu/dolphin/pull/6451 defaults MMU to on.

**Pull request closed without merging**